kbs/src/api_server.rs

```diff
@@ -0,0 +1,317 @@
+        let token_verifier = TokenVerifier::from_config(config.attestation_token.clone()).await?;
+        let policy_engine = PolicyEngine::new(&config.policy_engine).await?;
+        let admin_auth = Admin::try_from(config.admin.clone())?;
+
+        #[cfg(feature = "resource")]
+        let resource_storage =
+            crate::resource::ResourceStorage::try_from(config.repository.clone())?;
+
+        #[cfg(feature = "as")]
+        let attestation_service =
+            crate::attestation::AttestationService::new(config.attestation_service.clone()).await?;
+
+        Ok(Self {
+            config,
+            plugin_manager,
+            policy_engine,
+            admin_auth,
+            token_verifier,
+
+            #[cfg(feature = "resource")]
+            resource_storage,
+
+            #[cfg(feature = "as")]
+            attestation_service,
+        })
+    }
+
+    /// Start the HTTP server and serve API requests.
+    pub async fn serve(self) -> Result<()> {
+        info!(
+            "Starting HTTP{} server at {:?}",
+            if !self.config.http_server.insecure_http {
+                "S"
+            } else {
+                ""
+            },
+            self.config.http_server.sockets
+        );
+
+        let http_config = self.config.http_server.clone();
+        let http_server = HttpServer::new({
+            move || {
+                let api_server = self.clone();
+                App::new()
+                    .wrap(middleware::Logger::default())
+                    .app_data(web::Data::new(api_server))
+                    .service(
+                        web::resource([kbs_path!("{plugin}{sub_path:.*}")])
+                            .route(web::get().to(client))
+                            .route(web::post().to(client)),
+                    )
+                    .service(
+                        web::resource([kbs_path!("admin/{plugin}/{sub_path:.*}")])
+                            .route(web::get().to(admin))
+                            .route(web::post().to(admin)),
+                    )
+            }
+        });
+
+        if !http_config.insecure_http {
+            let tls_server = http_server
+                .bind_openssl(
+                    &http_config.sockets[..],
+                    crate::http::tls_config(&http_config)
+                        .map_err(|e| Error::HTTPSFailed { source: e })?,
+                )
+                .map_err(|e| Error::HTTPSFailed { source: e.into() })?;
+
+            return tls_server
+                .run()
+                .await
+                .map_err(|e| Error::ServerLaunchFailed { source: e.into() });
+        }
+
+        http_server
+            .bind(&http_config.sockets[..])
+            .map_err(|e| Error::HTTPFailed { source: e.into() })?
+            .run()
+            .await
+            .map_err(|e| Error::ServerLaunchFailed { source: e.into() })
+    }
+}
+
+/// Client APIs. /kbs/v0/XXX
+pub(crate) async fn client(
```

Comment:
Yes. You are right. I want to move the set_policy, set_resource APIs to `kbs/v0/admin/...`. Thus we could have a distinguished barrier from client api and admin api. Maybe we can relatively change the logic of kbs-client. This change will not influence anything on the guest-components side. wdyt?

Reply:
That's fine with me although I'm not sure it's entirely RESTful

Reply:
Nice catch for the RESTful design and client/admin APIs. Maybe I went to the wrong design way at first. We might only need one plugin system, and adds `admin_auth` to the `handle()`. This means it is the API that decides whether admin auth needs to be checked.

In this way, we can avoid extra `kbs/v0/admin/<plugin-name>`, but only have one API `kbs/v0/<plugin-name>`.

Wdyt?

Reply:
Yeah the problem with that is that you can't implement a plugin that has both `GET` / `POST` for the guest and `GET` / `POST` for an admin, which some people might want to do (although note that posting from the guest is a bit risky since the requests aren't encrypted). 

It looks like having an `/admin` route is somewhat common so maybe that's ok. Honestly I don't have a strong preference here.